bsip-0040.md

```diff
@@ -0,0 +1,155 @@
+
+The `asserts` is a list of `assert_objects` that are all together evaluated with `and` logic to evaluate a match
+```
+asserts = list of assert_object
+assert_object = {
+    argument, // target argument of the operationid
+    function, // functionid to do the assert 
+    data,     // stores data specific to the chosen function
+    state     // if this assert is statefull
+}
+```
+List of possible asserts are:
+
+| function        | data           | state  |
+| ------------- |:-------------:| -----:|
+| `any`      | [`list`, `of`, `allowed`, `values`] | stateless |
+| `range`      | [`min`, `max`] | stateless |
+| `length`      | [`min`, `max`] | stateless |
+| `limit` | [`max_cumsum`, `interval_in_sec`]      | [`current_cumsum`, `interval_began`] |
+| `limit_monthly` | [`max_cumsum`, `interval_in_months`]      | [`current_cumsum`, `interval_began`] |
+
+All asserts apply to `int`, `string` and `object_ids`. There is no type converstion, incompatible type means assert failure. Any more sophisticated data types are not included yet.
+
+#### `any`
+Stateless assert, all argument types. Value of argument must be equal to one of values in the list
+
+#### `range`
+Stateless assert, only `int` type arguments. Value must be inbetween `min` and `max`, inclusive. One of the bounds can be empty to indicate unlimited.
+
+#### `length`
+Stateless assert, only `string` type arguments. Length of string (or encrypted string in case of memo) must be within `min` and `max`, inclusive. One of the bounds can be empty to indicate unlimited.
+
+#### `limit`
+Statefull assert, only `int` type arguments. When the authority is created, `interval_began` is set to `valid_from` from its custom active authority and `max_cumsum` to `0`. Incoming operations are first tried to match all stateless asserts, 
+and if all passes continue with statefull asserts. If `now > interval_began + interval_in_sec`, then set `max_cumsum = 0` and set `interval_began = now`.
+The assert that needs to pass is now `current_cumsum + incoming value <= max_cumsum`. If all asserts are passed, update `current_cumsum = current_cumsum + incoming value` of all involved statefull asserts.
+
+#### `limit_monthly`
+Statefull assert, only `int` type arguments. Analogue to `limit`, but `interval_began` is initially set to `month(valid_from)` and set to `month(now)` on update, additionally the time assert is `month(now) >= interval_began + interval_in_months` (include logic for month overflow when year changes).
+
+### Simple Example
+Assume account A and B and some unrelated key K. Furthermore A has a custom active authority in the following way:
+```
+custom active authority = {
+    valid_from: 7.7.2018 00:00
+    valid_to: 8.7.2018 00:00
+    operationid: transfer,
+    authority: {
+       threshold: 1
+       key_auth: [K, 1]
+       account_auth: []
+    },
+    asserts: [
+        {
+            argument: to
+            function: any,
+            data: [B]
+        }
+    ]
+}
```

Comment:
Note: this is just an illustration of a possible serialization, not a specification of the serialized format.

Reply:
added as well